README.md
<a name="readme-top"></a>

<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/charliemalach/ValpoNav">
    <img src="./image/logo.png" alt="Logo" width="100" height="125">
  </a>

  <h3 align="center">Campus Information Prototype Application</h3>

  <p align="center">
    An exhaustive overview of the Campus Information Prototype Application
    <br />
  </p>
</div>

<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->

## About The Project

A student from Computing and Information Sciences (CIS) undertook to build a responsive web centered campus information mapping application as an independent study with Professor Nicholas S. Rosasco, DSc. This application would assist in navigation issues on campus and include information about the interior layouts for various buildings on campus. The CIS student opted for the use of HTML, CSS, and JavaScript to ensure that the application is fast, reliable, and easy to maintain. The responsive design of the application ensures that it is optimized for use on a variety of devices, including desktops, laptops, tablets, and smartphones. The CIS student also utilized the Leaflet.js Javascript library, as it is the leading open-source library which enables interactive maps that are also mobile friendly. The student worked closely together to define and prioritize the application's features, and then broke them down into smaller, manageable pieces of work called user stories, which were then prioritized and scheduled for development in short sprints, usually lasting one week. These choices enabled the system to be built in an “agile driven” collaborative approach, while still incorporating new innovations and ideas. Within the responsive application, users are able to search for specific buildings with autocompleting search functionality, or by browsing the organized ‘Navigation Menu’. The information provided for each building includes its name, description, and relevant photos. Extra features include a view change between Satellite and Map view, Current Geo location view, magnifying zoom functionality, and a reset view for the center of campus.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

### Built With

These are the tools I utilized throughout the implementation of my application:

- [![react][react-img]][react-url]
- [![expo][expo-img]][expo-url]
- [![json-server][json-img]][json-url]

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- GETTING STARTED -->

## Usage

This prototype is used as a 'pocket knife' for the Valparaiso University community. It offers many different features requested by potential users which assist with many different aspects of life on campus. Using this application, users are able to:

- Register and manage a Valpo User account
- Access the Founder's Menu
- Contact various Campus Resources
- Call Valparaiso University's Police Department with a 'Panic Button'
- Access the ValpoNav Campus Navigation application

![Product Name Screen Shot][screens]

Using ValpoNav, another application, users find much more information about buildings on campus. ValpoNav allows users to find a desired building which enables them to view inside to find more details about the building, including but not limited to:

- An image of the building
- A brief description of the building
- A link to the floorplans for the building

![Product Name Screen Shot][main-screenshot]
![Product Name Screen Shot][inside-screenshot]

See the [ValpoNav repository](https://github.com/charliemalach/ValpoNav) for a more indepth explanation of the application.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- ROADMAP -->

## Roadmap

- [x] Login / Register feature
- [x] DataVU Portal
- [x] BlackBoard Portal
- [x] Founder's Menu Accessibility
- [x] VUPD 'Panic Button'
- [x] Campus Resources
  - [x] Counseling Center
  - [x] Health Center
  - [x] Student Senate
- [x] ValpoNav Accessibility

See the [repository](https://github.com/charliemalach/ValpoNav) for a full list of proposed features (and known issues).

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- CONTACT -->

## Contact

Charlie Malachinski (Software Developer) - charlie.malachinski@valpo.edu

Project Link: [https://github.com/charliemalach/VU-Campus-Information-Prototype-Application](https://github.com/charliemalach/VU-Campus-Information-Prototype-Application)

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- ACKNOWLEDGMENTS -->

## Acknowledgments

Use this space to list resources you find helpful and would like to give credit to. I've included a few of my favorites to kick things off!

- [React](https://reactnative.dev/)
- [Expo Go](https://expo.dev/client)
- [Valparaiso University Facilities Management](https://www.valpo.edu/facilities-management/)
- [Valparaiso University Information Technology](https://www.valpo.edu/it)
- [ValpoNav Application](https://github.com/charliemalach/ValpoNav)
- [Professor Dushan Nikolovski, Clinical Associate Professor of Entrepreneurship, Director of Innovation](https://www.linkedin.com/in/dushannikolovski/)
- [Professor Nicholas S, Rosasco, DSc](https://www.valpo.edu/computing-information-sciences/faculty/nicholas-rosasco/)

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->

[screens]: ./image/screens.jpg
[main-screenshot]: ./image/main.jpg
[inside-screenshot]: ./image/inside.jpg
[expo-img]: https://camo.githubusercontent.com/f93c191872254e83a3c6664393dbbabbcca0de44a14f117d962a5718d4651379/68747470733a2f2f696d672e736869656c64732e696f2f7374617469632f76313f7374796c653d666f722d7468652d6261646765266d6573736167653d4578706f26636f6c6f723d303030303230266c6f676f3d4578706f266c6f676f436f6c6f723d464646464646266c6162656c3d
[expo-url]: https://expo.dev/client
[react-img]: https://camo.githubusercontent.com/67a01fa7cf337616274f39c070a11638f2e65720e414ef55b8dd3f9c2a803b2a/68747470733a2f2f696d672e736869656c64732e696f2f7374617469632f76313f7374796c653d666f722d7468652d6261646765266d6573736167653d526561637426636f6c6f723d323232323232266c6f676f3d5265616374266c6f676f436f6c6f723d363144414642266c6162656c3d
[react-url]: https://reactnative.dev/
[json-img]: https://camo.githubusercontent.com/9c2f1381d03b23626b66eb3372afe109aa0be6b50d1695c9ca939289290e39a7/68747470733a2f2f696d672e736869656c64732e696f2f7374617469632f76313f7374796c653d666f722d7468652d6261646765266d6573736167653d4a534f4e26636f6c6f723d303030303030266c6f676f3d4a534f4e266c6f676f436f6c6f723d464646464646266c6162656c3d
[json-url]: https://github.com/typicode/json-server


release-notes.txt
#----------------------------------------------------------------------------------------------------#

VU Campus Information Prototype Application - Version 0.0.4, 04/05/23

New Features: 
	- Ability to edit current user's profile avatar with 'Edit Profile' page

Known Issues: 
	- ValpoNav only works when on campus network (server hosting issue)

Expected Next Sprint: 
    - Fix ValpoNav connectivity issue (not sure if possible)

Links: 
	- GitHub Repository: https://github.com/charliemalach/VU-Campus-Information-Prototype-Application

#----------------------------------------------------------------------------------------------------#

VU Campus Information Prototype Application - Version 0.0.3, 03/24/23

New Features: 
	- Ability to edit current user's email, name, and username with 'Edit Profile' page
	- Ability to connect to ValpoNav for campus navigation 
    - Ability to edit current user details with 'Edit Profile' page

Known Issues: 
	- ValpoNav only works when on campus network (server hosting issue)
    - 'Edit Profile' is not saving user image to local databaes 

Expected Next Sprint: 
    - Ability to save images to user profile

Links: 
	- GitHub Repository: https://github.com/charliemalach/VU-Campus-Information-Prototype-Application

#----------------------------------------------------------------------------------------------------#

VU Campus Information Prototype Application - Version 0.0.2, 03/05/23

New Features: 
	- Ability to contact Campus Resources (Student Senate, Health Center, Counseling) 
	- User Login and Registration fully functional (including auto-increment user id) (JSON-server)
    - Ability to edit current user's username with 'Edit Profile' page

Known Issues: 
	- 'Edit Profile' is not saving user image to local databaes 

Expected Next Sprint: 
    - Ability to save images to user profile
    - Ability to use ValpoNav and navigate campus map 

Links: 
	- GitHub Repository: https://github.com/charliemalach/VU-Campus-Information-Prototype-Application

#----------------------------------------------------------------------------------------------------#

VU Campus Information Prototype Application - Version 0.0.1, 02/27/23

New Features: 
	- Ability to access Founder's menu 
	- Ability to contact VUPD 

Known Issues: 
	- Founder's menu is displayed with my personal website instead of in-application view 
    - VUPD 'Panic Button' calls my phone number instead of VUPD to prevent false alerts
    - Local Database does not save the new information
    - Local Database is not working for accessing user information 

Expected Next Sprint: 
    - Ability to contact Campus Resources

Links: 
	- GitHub Repository: https://github.com/charliemalach/VU-Campus-Information-Prototype-Application

#----------------------------------------------------------------------------------------------------#

VU Campus Information Prototype Application - Version 0.0.0, 02/16/23

New Features: 
	- Created a Login and Register account page
	- Ability for user to navigate to Home page 
    - Ability for user to connect to Blackboard
    - Ability for user to connect to DataVU

Known Issues: 
	- Local Database does not save the new information
    - Local Database is not working for accessing user information 

Expected Next Sprint: 
	- Ability to reach Founder's Menu 
    - Ability to contact Campus Resources 
    - Ability to contact VUPD 

Links: 
	- GitHub Repository: https://github.com/charliemalach/VU-Campus-Information-Prototype-Application

#----------------------------------------------------------------------------------------------------#

Login.js
import React, { useState } from 'react';
import { View, Text, TextInput, TouchableOpacity, Image, Alert, StyleSheet, StatusBar } from 'react-native';
import { useNavigation } from '@react-navigation/native';


const API_URL = 'https://d773-152-228-20-39.ngrok.io'; //change this link for the user db 
const defaultImage = "https://cdn.greatnews.life/wp-content/uploads/images/vu-logo.jpg";

let loggedName = 'name';
let loggedUser = 'username';
let loggedImage = defaultImage;
let loggedEmail = 'email';

const Login = () => {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [image, setImage] = useState('');
  const [nextId, setNextId] = useState(1);
  const [isLogin, setIsLogin] = useState(true);
  const navigation = useNavigation();
  
  const handleLogin = async () => {
    try {
      const response = await fetch(`${API_URL}/users?username=${username}&password=${password}`, {method: "GET"});
      const [user] = await response.json();
      if (user) {
        console.log("User logged in:", user);
        loggedName = user.name;
        loggedUser = user.username;
        loggedImage = user.image;
        loggedEmail = user.email;
        // handle successful login, e.g. navigate to home screen
        navigation.navigate('Home', { username: loggedUser });
        
        console.log(loggedUser);
        setName('');
        setUsername('');
        setPassword('');
        setImage('');
        setEmail('');
        setIsLogin(true);
        
        StatusBar.setHidden(true);
        return loggedUser, loggedImage, loggedEmail, loggedName;
      } else {
        console.error("Invalid credentials");
        // handle failed login, e.g. display error message to user
      }
    } catch (error) {
      console.error(error);
      // handle error, e.g. display error message to user
    }
    return loggedUser;
  };
  
  const handleRegister = async () => {
    try {
      const response = await fetch(`${API_URL}/users?username=${username}`);
      const [user] = await response.json();
      if (user) {
        console.error("User already exists");
        Alert.alert("ERROR: User already exists. You have not registered.");
        // handle failed registration, e.g. display error message to user
      } else {
        setImage(defaultImage)
        const newUser = { id: nextId, name, username, password, image, email }; //for some reason messing with the space after email allows post requests ..?
        const response = await fetch(`${API_URL}/users`, {
          method: 'POST',
          headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
          },
          body: JSON.stringify(newUser)
        });
        const data = await response.text();
        console.log("User registered:", data);
        setNextId(nextId + 1); // increment the nextId for the next user
        setImage(defaultImage)
        Alert.alert("You have successfully registered");
        // handle successful registration, e.g. display success message to user
      }
    } catch (error) {
      console.error(error);
      // handle error, e.g. display error message to user
    }
  };


  const styles = StyleSheet.create({
    container: {
      flex: 1,
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: '#ffcc00',
    },
    title: {
      fontSize: 24,
      marginBottom: 24,
    },
    input: {
      width: '80%',
      padding: 12,
      marginBottom: 16,
      backgroundColor: '#eee',
      borderRadius: 4,
    },
    button: {
      width: '80%',
      padding: 12,
      backgroundColor: '#613318',
      borderRadius: 4,
      alignItems: 'center',
    },
    buttonText: {
      color: '#fff',
      fontSize: 18,
    },
    switchText: {
      marginTop: 24,
      color: '#4287f5',
    },
    tinyLogo: {
      width: 100,
      height: 100,
      resizeMode: 'contain',
    },
});

  return (
    <View style={styles.container}>
  
      <Image
        style={styles.tinyLogo}
        source={require('./assets/shield.png')}
      />

      <Text style={styles.title}>{isLogin ? 'Login' : 'Register'}</Text>

      { !isLogin &&
        <TextInput
          style={styles.input}
          placeholder="Name"
          value={name}
          secureTextEntry={false}
          onChangeText={setName}
        />
      }

      { !isLogin &&
        <TextInput
          style={styles.input}
          placeholder="Email"
          value={email}
          secureTextEntry={false}
          onChangeText={setEmail}
        />
      }

      <TextInput
        style={styles.input}
        placeholder="Username"
        value={username}
        secureTextEntry={false}
        onChangeText={setUsername}
      />

      <TextInput
        style={styles.input}
        placeholder="Password"
        value={password}
        secureTextEntry={true}
        onChangeText={setPassword}
      />

      <TouchableOpacity style={styles.button} onPress={isLogin ? handleLogin : handleRegister}>
        <Text style={styles.buttonText}>{isLogin ? 'Login' : 'Register'}</Text>
      </TouchableOpacity>

      <TouchableOpacity onPress={() => setIsLogin(!isLogin)}>
        <Text style={styles.switchText}>
          {isLogin ? 'New user? Register here' : 'Already have an account? Login here'}
        </Text>
      </TouchableOpacity>
    </View>
  );
}

export {API_URL};
export {loggedName};
export {loggedUser};
export {loggedImage};
export {loggedEmail};
export default Login